### PR TITLE
[MNG-7236] The DefaultPluginVersionResolver should cache results for …

### PR DESCRIPTION
**Pull request by gnodet**
…the session

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [X] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

File: maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java [hunk 1]
```diff
@@ -384,6 +401,36 @@ private PluginVersionResult resolveFromProject( PluginVersionRequest request, Li
         return null;
     }
 
+    @SuppressWarnings( "unchecked" )
+    private Map<String, PluginVersionResult> getCache( SessionData data )
+    {
+        Map<String, PluginVersionResult> cache = ( Map<String, PluginVersionResult> ) data.get( CACHE_KEY );
+        while ( cache == null )
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java:408]: Why not use `computeIfAbsent?

Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java:408]: That would be awesome, but `SessionData` interface is not a `Map` unfortunately !...

Comment (michael-o) [reply maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java:408]: The problem is that this spot is not atomic,  is it?

File: maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java [hunk 2]
```diff
@@ -87,22 +92,37 @@
     public PluginVersionResult resolve( PluginVersionRequest request )
         throws PluginVersionResolutionException
     {
+        ConcurrentMap<String, PluginVersionResult> cache = getCache( request.getRepositorySession().getData() );
+        String key = getKey( request );
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java:96]: You don't need the key here, do you? Only when `result` is `null`. I would pull it down where it belongs.

Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java]: @michael-o you're right, this is fixed.  I've also aligned the PR on `master` to use a `ConcurrentMap` too (this was required to support JDK 1.7).

**Review by michael-o — APPROVED** [review]
This looks reasonable to me.
@rfscholte Any objections?

**Review by rfscholte — APPROVED** [review]

**Comment by michael-o** [comment]
@gnodet Please go ahead, rebase and merge into master and 3.8.x.

**Comment by jira-importer** [comment]
Resolve #8209

**Pull request merged**